### PR TITLE
CPLGetPhysicalRAM(): take into account current cgroup (v1 and v2)

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

File: port/cpl_vsisimple.cpp [hunk 1]
```diff
@@ -1371,20 +1371,57 @@ GIntBig CPLGetPhysicalRAM(void)
...
     {
-        char szBuffer[32];
-        const int nRead =
-            static_cast<int>(fread(szBuffer, 1, sizeof(szBuffer) - 1, f));
-        szBuffer[nRead] = 0;
+        FILE *f = fopen("/proc/self/cgroup", "rb");
+        char szLine[256];
+        // Find line like "6:memory:/user.slice/user-1000.slice/user@1000.service"
+        // and store "/user.slice/user-1000.slice/user@1000.service" in szMemoryPath
+        while (fgets(szLine, sizeof(szLine) - 1, f))
```
Comment (lnicola) [inline port/cpl_vsisimple.cpp:1382]:
```suggestion
        while (fgets(szLine, sizeof(szLine), f))
```

`fgets` reads one character less than the size passed in, then adds a null.

Comment (rouault) [reply port/cpl_vsisimple.cpp]: thanks. fixed

File: port/cpl_vsisimple.cpp [hunk 2]
```diff
@@ -1371,20 +1371,57 @@ GIntBig CPLGetPhysicalRAM(void)
...
-        nVal =
-            static_cast<GIntBig>(std::min(static_cast<GUIntBig>(nVal), nLimit));
     }
+    if (szMemoryPath[0])
+    {
+        char szFilename[256 + 64];
+        // Read memory.limit_in_byte in the whole szMemoryPath hierarchy
+        while (true)
+        {
+            snprintf(szFilename, sizeof(szFilename),
```
Comment (lnicola) [inline port/cpl_vsisimple.cpp:1402]: I think this only works with the `cgroupv1` hierarchy, but most distros have switched or are switching to `cgroupv2` by now. See https://chrisdown.name/talks/cgroupv2/cgroupv2-fosdem.pdf for more details.

On my distro, I can spawn a new cgroup with `systemd-run --scope -p MemoryMax=250M -p MemoryHigh=200M bash`, then:

```sh
cat /proc/self/cgroup 
0::/system.slice/run-u4260.scope
$ ls /sys/fs/cgroup/system.slice/run-u4260.scope
cgroup.controllers      cgroup.threads         cpuset.mems.effective     hugetlb.1GB.numa_stat     io.bfq.weight   memory.events        memory.reclaim        misc.max
cgroup.events           cgroup.type            cpu.stat                  hugetlb.1GB.rsvd.current  io.latency      memory.events.local  memory.stat           pids.current
cgroup.freeze           cpu.idle               cpu.uclamp.max            hugetlb.1GB.rsvd.max      io.low          memory.high          memory.swap.current   pids.events
cgroup.kill             cpu.max                cpu.uclamp.min            hugetlb.2MB.current       io.max          memory.low           memory.swap.events    pids.max
cgroup.max.depth        cpu.max.burst          cpu.weight                hugetlb.2MB.events        io.pressure     memory.max           memory.swap.high      pids.peak
cgroup.max.descendants  cpu.pressure           cpu.weight.nice           hugetlb.2MB.events.local  io.prio.class   memory.min           memory.swap.max       rdma.current
cgroup.pressure         cpuset.cpus            hugetlb.1GB.current       hugetlb.2MB.max           io.stat         memory.numa_stat     memory.zswap.current  rdma.max
cgroup.procs            cpuset.cpus.effective  hugetlb.1GB.events        hugetlb.2MB.numa_stat     io.weight       memory.oom.group     memory.zswap.max
cgroup.stat             cpuset.cpus.partition  hugetlb.1GB.events.local  hugetlb.2MB.rsvd.current  irq.pressure    memory.peak          misc.current
cgroup.subtree_control  cpuset.mems            hugetlb.1GB.max           hugetlb.2MB.rsvd.max      memory.current  memory.pressure      misc.events
$ cat /sys/fs/cgroup/system.slice/run-u4260.scope/memory.max 
262144000
```

(basically, the resources and cgroups are switched in the hierarchy).

Comment (rouault) [reply port/cpl_vsisimple.cpp:1402]: I'm on Ubuntu 20.04.... Thanks for the info!

Comment (lnicola) [reply port/cpl_vsisimple.cpp:1402]: Just tested on Ubuntu 22.04, it's using the same (v2) hierarchy.

Comment (rouault) [reply port/cpl_vsisimple.cpp]: support for cgroup v1 (tested on 20.04) and v2 (tested on 22.04) added

File: port/cpl_vsisimple.cpp [hunk 3]
```diff
@@ -1371,20 +1372,107 @@ GIntBig CPLGetPhysicalRAM(void)
...
+                    nVal = static_cast<GIntBig>(
+                        std::min(static_cast<GUIntBig>(nVal), nLimit));
+                }
+                char *pszLastSlash = strrchr(szGroupName, '/');
+                if (!pszLastSlash)
+                    break;
+                *pszLastSlash = '\0';
+            }
+        }
+        else
```
Comment (lnicola) [inline port/cpl_vsisimple.cpp:1439]: Nit: might be worth trying V2 first, since it's going to be the most used path in the future (if not already).

**Comment by lnicola** [comment]
I haven't tested, but `CPLGetNumCPUs` might have a similar problem.

File: port/cpl_vsisimple.cpp (hunk 3 above)
Comment (rouault) [reply port/cpl_vsisimple.cpp:1439]: the logic (possibly flawed) is that if there is a ":memory:" line in "/proc/self/cgroup" then the V1 code path is taken. And if there's only a "0::" line, then V2 code path is taken.

Comment (lnicola) [reply port/cpl_vsisimple.cpp:1439]: That's right, sorry.

**Comment by rouault** [comment]
> I haven't tested, but `CPLGetNumCPUs` might have a similar problem.

The following works on Ubuntu 22.04:
```
$ sudo systemd-run --scope -p AllowedCPUs=0,1,2 autotest/cpp/gdallimits
Running scope as unit: run-rc20e757cc8b944cb8fa2651424bda12e.scope
CPLGetNumCPUs = 3
CPLGetUsablePhysicalRAM = 3 GB
```

**Pull request merged**